src/platform/backends/hyperv_api/hcs_virtual_machine.cpp

```
@@ -0,0 +1,748 @@
+                                           create_result.code};
+    }
+
+    // Grant access to the VHDX and the cloud-init ISO files.
+    for (const auto& scsi : create_compute_system_params.scsi_devices)
+    {
+        if (scsi.type == hcs::HcsScsiDeviceType::VirtualDisk())
+        {
+            std::vector<std::filesystem::path> lineage{};
+            if (VirtDisk().list_virtual_disk_chain(scsi.path.get(), lineage))
+            {
+                grant_access_to_paths({lineage.begin(), lineage.end()});
+            }
+        }
+        else
+        {
+            grant_access_to_paths({scsi.path.get()});
+        }
+    }
+
+    return true;
+}
+
+void HCSVirtualMachine::set_state(hcs::ComputeSystemState compute_system_state)
+{
+    mpl::debug(log_category,
+               "set_state() -> VM `{}` HCS state `{}`",
+               get_name(),
+               compute_system_state);
+
+    const auto prev_state = state;
+    switch (compute_system_state)
+    {
+    case hcs::ComputeSystemState::created:
+        state = State::off;
+        break;
+    case hcs::ComputeSystemState::paused:
+        state = State::suspended;
+        break;
+    case hcs::ComputeSystemState::running:
+        state = State::running;
+        break;
+    case hcs::ComputeSystemState::saved_as_template:
+    case hcs::ComputeSystemState::stopped:
+        state = State::stopped;
+        break;
+    case hcs::ComputeSystemState::unknown:
+        state = State::unknown;
+        break;
+    }
+
+    if (state == prev_state)
+        return;
+
+    mpl::info(log_category,
+              "set_state() > VM {} state changed from {} to {}",
+              get_name(),
+              prev_state,
+              state);
+}
+
+void HCSVirtualMachine::start()
+{
+    mpl::debug(log_category, "start() -> Starting VM `{}`, current state {}", get_name(), state);
+
+    // Create the compute system, if not created yet.
+    if (maybe_create_compute_system())
```

Hmm, why? This is not what the old backend did. If the VM doesn't exist at this point, something went wrong or someone removed it, correct? If so, better fail early no? One VM is not the same as the next.

---

See my previous comment about when does a VM gets recreated.